### PR TITLE
Use mps device on macOS

### LINKED ISSUE #79: macOS M1 (Apple Silicon) Support?
Would be nice

### PR DESCRIPTION
**Pull request by akx**
Fixes #79.

Works fine on my Macbook, inferences at around 1.40it/s (c.f. 4.85s/it on `cpu`).

### CODE REVIEW

**Review by TheMattBin — APPROVED** [review]

**Comment by scf4** [comment]
Wjhat Macbook do you have, for reference?

Also is this PR effectively slowing things down by 3.5x by default?

**Comment by TheMattBin** [comment]
so weird that running on Mac slowing down the process

**Comment by scf4** [comment]
it’s optimized for cuda. pytorch mps is often slower than cpu

**Comment by akx** [comment]
> Which Macbook do you have, for reference?

The silicon is M2 Max, 64 GB.

> Also is this PR effectively slowing things down by 3.5x by default?

No. 4.85 seconds per iteration is 0.2062 iterations per second. The `mps` speed, 1.40 iterations per second, is 6.7x faster.

> so weird that running on Mac slowing down the process

Well, prior to this the repo only detects CUDA, which never exists on Apple Silicon, for acceleration. You'd get the same speed (or lack thereof) on a machine with an AMD GPU, since it wouldn't be detected, and it'd use CPU only.

The downside with `mps` is that some operations and data formats (and combinations thereof) don't work, but as said, this empirically does work.

**Comment by scf4** [comment]
Hey sorry you’re right. Misread it/sec and s/it

File: stable_audio_tools/interface/gradio.py [hunk 1]
```diff
@@ -665,7 +667,15 @@ def create_ui(model_config_path=None, ckpt_path=None, pretrained_name=None, pret
     else:
         model_config = None
 
-    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    if platform.system() == "Darwin":
```
Comment (zqevans) [inline stable_audio_tools/interface/gradio.py:670]: I believe using the `torch.backend.mps.is_available()` call is more appropriate here, docs here: https://pytorch.org/docs/stable/notes/mps.html (not sure what PyTorch version is needed for this)

Comment (akx) [reply stable_audio_tools/interface/gradio.py]: Sure. Amended to have a robust check for that.

**Comment by bghira** [comment]
> You'd get the same speed (or lack thereof) on a machine with an AMD GPU, since it wouldn't be detected

AMD ROCm PyTorch masquerades as a cuda backend. it's plug-and-play.

File: stable_audio_tools/interface/gradio.py [hunk 2]
```diff
@@ -665,7 +667,21 @@ def create_ui(model_config_path=None, ckpt_path=None, pretrained_name=None, pret
...
+        has_mps = False
+
+    if has_mps:
+        device = torch.device("mps")
+    elif torch.cuda.is_available():
+        device = torch.device("cuda")
+    else:
+        device = torch.device("cpu")
+
+    print("Using device:", device)
```
Comment (zqevans) [inline stable_audio_tools/interface/gradio.py:683]: Remove this print and this is probably good to go

Comment (akx) [reply stable_audio_tools/interface/gradio.py:683]: I think it's very useful to see at a glance which device is being used. For instance, seeing `Using device: cpu` when you're on an ostensibly CUDA machine...

**Comment by andrewgazelka** [comment]
Is there something blocking this being merged?

**Comment by carlthome** [comment]
> Is there something blocking this being merged?

I guess https://github.com/Stability-AI/stable-audio-tools/pull/82#discussion_r1633928547 is the remaining action. Would be great to have MPS support in the Gradio app! 🎉

**Comment by akx** [comment]
> I guess [#82 (comment)](https://github.com/Stability-AI/stable-audio-tools/pull/82#discussion_r1633928547) is the remaining action. 

Yeah, I'm waiting for @zqevans to reply on that – I think it'd be very useful to know which device is in use.

**Pull request merged**